sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -63,16 +74,345 @@ abstract class GpuShuffleBlockResolverBase(
   override def stop(): Unit = wrapped.stop()
 }
 
+/**
+ * The `ShuffleWriteMetricsReporter` is based on accumulators, which are not thread safe.
+ * This class is a thin wrapper that adds synchronization, since these metrics will be written
+ * by multiple threads.
+ * @param wrapped
+ */
+class ThreadSafeShuffleWriteMetricsReporter(wrapped: ShuffleWriteMetricsReporter)
+  extends ShuffleWriteMetrics {
+  override private[spark] def incBytesWritten(v: Long): Unit = synchronized {
+    wrapped.incBytesWritten(v)
+  }
+  override private[spark] def incRecordsWritten(v: Long): Unit = synchronized {
+    wrapped.incRecordsWritten(v)
+  }
+  override private[spark] def incWriteTime(v: Long): Unit = synchronized {
+    wrapped.incWriteTime(v)
+  }
+  override private[spark] def decBytesWritten(v: Long): Unit = synchronized {
+    wrapped.decBytesWritten(v)
+  }
+  override private[spark] def decRecordsWritten(v: Long): Unit = synchronized {
+    wrapped.decRecordsWritten(v)
+  }
+}
 
 object RapidsShuffleInternalManagerBase extends Logging {
   def unwrapHandle(handle: ShuffleHandle): ShuffleHandle = handle match {
     case gh: GpuShuffleHandle[_, _] => gh.wrapped
     case other => other
   }
+
+  /**
+   * A single-thread + queue used to compute a series of tasks serially.
+   * Used to compose a map of slots that can run independent of each other.
+   */
+  case class Task(body: () => Unit, handleException: Throwable => Unit)
+  private class Slot(i: Int) {
+    val tasks = new LinkedBlockingQueue[Task]()
+    val p = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
```

Comment:
Why is there a single-threaded threadpool per slot rather than a threadpool to execute slots in?

Reply:
So I am trying to handle the case where a task may have multiple batches, where each batch would contain a partition for reducer X.

Given that, I would like to place all tasks for reducer X in a single slot, hence the one-to-one relationship here.

So when I see a writer for a partition, I pick a slot and put it in its queue, so prior writes for that partition are not getting clobbered.